### ISSUE TITLE
[err] event.c:1673

### ISSUE THREAD

**Issue by iwroldtan** [libevent/libevent#513]
i write a http server with libevent API. 
A error happens after period of time.
[err] event.c:1673: Assertion (evcb->evcb_flags & EVLIST_FINALIZING) failed in event_process_active_single_queue

**Comment by azat** [comment]
Hi @iwroldtan 

Can you post a simple reproducer? (lesser lines - better)

**Comment by iwroldtan** [comment]
void httpserver_genericHandler(struct evhttp_request *req, void *arg)
{
	// Takes ownership of the request object.
	evhttp_request_own(req);

	EnterCriticalSection(&g_csThreadReq);
	if (reqQueue.size() <= MAX_REQ_SIZE)
	{
		reqQueue.push(req);
		LeaveCriticalSection(&g_csThreadReq);
	}
	else
	{
		LeaveCriticalSection(&g_csThreadReq);
		evhttp_send_error(req, 513, "the server is busy");
	}
}

i guess the error happen at when i own the req. but if i evhttp_request_free(req), another error will happen.(access violation reading 0xfeeefeee)

**Comment by azat** [comment]
Yep it is valid after `evhttp_request_own()`, but I remember there was some bugs with it, which version of libevent do you use? Can you try latest 2.1.8?

**Comment by azat** [comment]
Bugs like this one b0d3964ff5cf26a0d50ebef72fef177f75e9a025

**Comment by iwroldtan** [comment]
yes, I use the latest libevent-2.1.8-stable.
I want to use threadpool to manage the request. so I own all request and then I use other thread to deal with request. after more than 27000 request, the error happens.

I look up the libevent-2.1.8-stable source file.
My error is not the request was free by auto, but I cannot free my owned request logical.
failed in event_process_active_single_queue.
In short, I guess the queue is full result in the error.

**Comment by iwroldtan** [comment]
I want to know waht will cause the error, [err] event.c:1673: Assertion (evcb->evcb_flags & EVLIST_FINALIZING) failed in event_process_active_single_queue. Thank you!

**Comment by azat** [comment]
This assert means some corruption of event structure (at least I don't see how this can happen otherwise). So can you compiler your program with `-fsanitize=address -fsanitize=thread` or run it under valgrind with `drd/helgrind` for possible races?

**Comment by iwroldtan** [comment]
thank you for your suggestion. 
well, I have another question, I free the request by evhttp_request_free() in the complete_cb function, but failed. Acess violation resding location 0x0000000. This error break at  event_base_dispatch(server->base);
how can I deal with it?

**Comment by azat** [comment]
> Acess violation resding location 0x0000000.

s/resding/reseting/ ?

So it seems that you are trying to free NULL request, you can add assert before it (of course your should compile without `NDEBUG` defined otherwise assert will compile out)l.

And also what I suggested earlier in this issue will help you to find all this places (since `-fsanitize=address` is the address sanitizer, like `valgrind` `memcheck`)

> This error break at event_base_dispatch(server->base);

What do you mean by this?

**Comment by iwroldtan** [comment]
No, the req is not NULL. After I free the request, the event_base_loop just read something in the request. I do not know. So it shows "Acess violation resding location 0x0000000."
I mean the error is in event_base_loop, I guess.

**Comment by azat** [comment]
Hm, this is strange, but I just looked through the http code and it seems that you cannot free request from chunked callback, since it is used after callback is called, but this is client and this is not your case.

Maybe some sort of race condition or corruption? Again can you run your program under sanitizers to have more real issues than guesses?
